toolkit/tools/pkg/imagecustomizerlib/partitionutils.go

```diff
@@ -347,24 +389,207 @@ func findPartition(mountIdType imagecustomizerapi.MountIdentifierType, mountId s
+		return partition.Uuid == verityPartitionId, nil
+	default:
+		return false, fmt.Errorf("unsupported mount identifier type: (%s)", mountIdType)
+	}
+}
+
+func extractKernelCmdline(partitions []diskutils.PartitionInfo, buildDir string) ([]grubConfigLinuxArg, error) {
+	espPartition, err := findSystemBootPartition(partitions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find ESP partition: %w", err)
+	}
+
+	bootPartition, err := findBootPartitionFromEsp(espPartition, partitions, buildDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find boot partition: %w", err)
+	}
+
+	tmpDirEsp := filepath.Join(buildDir, tmpEspPartitionDirName)
+	espPartitionMount, err := safemount.NewMount(espPartition.Path, tmpDirEsp, espPartition.FileSystemType, 0, "", true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount ESP partition (%s):\n%w", espPartition.Path, err)
+	}
+	defer espPartitionMount.Close()
+
+	cmdline, err := extractKernelCmdlineFromUki(tmpDirEsp, buildDir)
+	if err == nil {
+		return cmdline, nil
+	} else if !errors.Is(err, os.ErrNotExist) {
+		return nil, err
+	}
+
+	tmpDirBoot := filepath.Join(buildDir, tmpBootPartitionDirName)
+	bootPartitionMount, err := safemount.NewMount(bootPartition.Path, tmpDirBoot, bootPartition.FileSystemType, unix.MS_RDONLY, "", true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount boot partition (%s):\n%w", bootPartition.Path, err)
+	}
+	defer bootPartitionMount.Close()
+
+	cmdline, err = extractKernelCmdlineFromGrub(tmpDirBoot)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract kernel arguments from grub.cfg:\n%w", err)
+	}
+
+	err = bootPartitionMount.CleanClose()
+	if err != nil {
+		return nil, fmt.Errorf("failed to close bootPartitionMount:\n%w", err)
+	}
+
+	err = espPartitionMount.CleanClose()
+	if err != nil {
+		return nil, fmt.Errorf("failed to close espPartitionMount:\n%w", err)
+	}
+
+	return cmdline, nil
+}
+
+func extractKernelCmdlineFromUki(tmpDirEsp, buildDir string) ([]grubConfigLinuxArg, error) {
+	espLinuxPath := filepath.Join(tmpDirEsp, UkiOutputDir)
+	ukiFiles, err := filepath.Glob(filepath.Join(espLinuxPath, "vmlinuz-*.efi"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to search for UKI images in ESP partition:\n%w", err)
+	}
+
+	if len(ukiFiles) == 0 {
+		return nil, os.ErrNotExist
+	}
+
+	cmdlinePath := filepath.Join(buildDir, "cmdline.txt")
+	_, _, err = shell.Execute("objcopy", "--dump-section", ".cmdline="+cmdlinePath, ukiFiles[0])
+	if err != nil {
+		return nil, fmt.Errorf("failed to dump kernel cmdline args from UKI:\n%w", err)
+	}
+
+	cmdlineContent, err := os.ReadFile(cmdlinePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read kernel cmdline args from dumped file:\n%w", err)
+	}
+
+	tokens, err := grub.TokenizeConfig(string(cmdlineContent))
+	if err != nil {
+		return nil, fmt.Errorf("failed to tokenize kernel command-line from UKI: %w", err)
+	}
+
+	args, err := ParseCommandLineArgs(tokens)
```

Comment:
You should use `getLinuxCommandLineArgs` instead of `ParseCommandLineArgs`.

`getLinuxCommandLineArgs` searches for the `linux` command and then parse the kernel args. `ParseCommandLineArgs` only parses the kernel args.

Reply:
There will not be able to search `linux <kernel>` from UKI, due to trim the first two strings (linux and its kernel) when building the UKI.

Reply:
Oh, sorry. I misread the code.

You are already using `getLinuxCommandLineArgs` in `extractKernelCmdlineFromGrub`. And you are correctly using `ParseCommandLineArgs` in `extractKernelCmdlineFromUki`.